packagedb/api.py

```diff
@@ -709,6 +509,196 @@ class PackageSetViewSet(viewsets.ReadOnlyModelViewSet):
     queryset = PackageSet.objects.prefetch_related('packages')
     serializer_class = PackageSetAPISerializer
 
+class CollectViewSet(viewsets.ViewSet):
+    """
+    Return Package data for the purl passed in the `purl` query parameter.
+
+    If the package does not exist, we will fetch the Package data and return
+    it in the same request.
+    
+    **Note:** Use `Index packages` for bulk indexing of packages; use `Reindex packages`
+    for bulk reindexing of existing packages.
+    """
+
+    def list(self, request, format=None):
+        purl = request.query_params.get('purl')
+
+        # validate purl
+        try:
+            package_url = PackageURL.from_string(purl)
+        except ValueError as e:
+            message = {
+                'status': f'purl validation error: {e}'
+            }
+            return Response(message, status=status.HTTP_400_BAD_REQUEST)
+
+        lookups = purl_to_lookups(purl)
+        packages = Package.objects.filter(**lookups)
+        if packages.count() == 0:
+            try:
+                errors = priority_router.process(purl)
+            except NoRouteAvailable:
+                message = {
+                    'status': f'cannot fetch Package data for {purl}: no available handler'
+                }
+                return Response(message, status=status.HTTP_400_BAD_REQUEST)
+
+            lookups = purl_to_lookups(purl)
+            packages = Package.objects.filter(**lookups)
+            if packages.count() == 0:
+                message = {}
+                if errors:
+                    message = {
+                        'status': f'error(s) occured when fetching metadata for {purl}: {errors}'
+                    }
+                return Response(message)
+
+        serializer = PackageAPISerializer(packages, many=True, context={'request': request})
+        return Response(serializer.data)
+    
+    @action(detail=False, methods=['post'])
+    def index_packages(self, request, *args, **kwargs):
```

Comment:
@keshav-space @pombredanne 

What do you guys think about merging `index_packages` and `reindex_packages` into a single method?

Reply:
Yes, we can do that.

A typical request to the merged endpoint would look like this:
```json
{
    "packages": [
        {
            "purl": "pkg:npm/less@1.0.32",
            "vers": null
        },
        {
            "purl": "pkg:npm/less",
            "vers": "vers:npm/>=1.1.0|<=1.1.4"
        },
        {
            "purl": "pkg:npm/foobar",
            "vers": null
        }
    ]
    "reindex":  true,
    "reindex_set": true,
}
```
@JonoYang What should be the name of the merged endpoint?

@JonoYang Also I am wondering, suppose purls are sent with the `reindex` flag set to true, and if certain purls are not already available on PurlDB, should we proceed to index those packages, or should we continue returning them as `nonexistent_packages`, as we are currently doing in our `reindex_packages` endpoint?